File: frontend/src/lib/languages.ts
/** Language catalogue — mirrors backend/apps/accounts/languages.py.
 *
 * Three different language questions live in this app and must not be
 * conflated:
 *  - `ui_language` — what the interface is written in (en/fa only, that's what
 *    we've translated).
 *  - `learning_languages` — what the user wants to learn.
 *  - `known_languages` — what they already understand.
 *
 * The last two drive which reels reach them: a reel teaches a *target* language
 * *in* a *base* language, and only matches when the target is something they're
 * learning and the base something they already read. Someone learning German
 * who only reads English shouldn't be shown German explained in Persian.
 */

export type LanguageCode =
  | "de" | "en" | "fa" | "ar" | "tr" | "fr" | "es" | "it" | "ru" | "nl";

export type Language = {
  code: LanguageCode;
  /** English name, for the en UI. */
  name: string;
  /** How speakers write it themselves — what someone picking their own
   *  language should see. */
  endonym: string;
};

export const LANGUAGES: Language[] = [
  { code: "de", name: "German", endonym: "Deutsch" },
  { code: "en", name: "English", endonym: "English" },
  { code: "fa", name: "Persian", endonym: "فارسی" },
  { code: "ar", name: "Arabic", endonym: "العربية" },
  { code: "tr", name: "Turkish", endonym: "Türkçe" },
  { code: "fr", name: "French", endonym: "Français" },
  { code: "es", name: "Spanish", endonym: "Español" },
  { code: "it", name: "Italian", endonym: "Italiano" },
  { code: "ru", name: "Russian", endonym: "Русский" },
  { code: "nl", name: "Dutch", endonym: "Nederlands" },
];

export function languageLabel(code: string): string {
  const lang = LANGUAGES.find((l) => l.code === code);
  if (!lang) return code;
  return lang.name === lang.endonym ? lang.name : `${lang.name} (${lang.endonym})`;
}

/** Has the user told us what they're learning? The reels feed asks when this is
 *  false rather than guessing — one extra question beats a feed of videos they
 *  can't follow. */
export function hasLanguagePrefs(user: {
  learning_languages?: string[] | null;
} | null): boolean {
  return !!user?.learning_languages?.length;
}

File: backend/apps/accounts/languages.py
"""The language catalogue — the single source of truth for language codes.

Two different questions are asked about a user, and they must not be confused:

  * `ui_language` — what the *interface* is written in. Only en/fa, because
    that's what we've translated.
  * `learning_languages` / `known_languages` — what the user wants to learn and
    what they already understand. These drive content matching, notably which
    reels reach them (see apps.reels).

Content carries the mirror image of the same pair: a reel has a
**target language** (what it teaches) and a **base language** (what it's
explained in). An Instagram account teaching German to Persian speakers is
target=de, base=fa. A reel matches a user when the target is something they're
learning and the base is something they already understand.

Keep in sync with the frontend mirror in frontend/src/lib/languages.ts.
"""

# code -> (English name, endonym). The endonym matters: someone picking their
# own language should see it written the way they write it.
LANGUAGES = {
    "de": ("German", "Deutsch"),
    "en": ("English", "English"),
    "fa": ("Persian", "فارسی"),
    "ar": ("Arabic", "العربية"),
    "tr": ("Turkish", "Türkçe"),
    "fr": ("French", "Français"),
    "es": ("Spanish", "Español"),
    "it": ("Italian", "Italiano"),
    "ru": ("Russian", "Русский"),
    "nl": ("Dutch", "Nederlands"),
}

LANGUAGE_CODES = list(LANGUAGES)
LANGUAGE_CHOICES = [(code, names[0]) for code, names in LANGUAGES.items()]

# Content that teaches a language *in* that language — a German-only reel with
# no translation. Stored as "" rather than a code: it isn't a language the user
# has to already know, so it reaches every learner of the target language.
IMMERSIVE = ""


def label(code: str) -> str:
    """"German (Deutsch)", or the raw code if we don't know it."""
    names = LANGUAGES.get(code)
    if not names:
        return code
    english, endonym = names
    return english if english == endonym else f"{english} ({endonym})"


def clean_codes(values) -> list[str]:
    """Normalise a client-supplied list: known codes only, lower-cased,
    de-duplicated, order preserved (the first entry is the user's primary)."""
    if not isinstance(values, (list, tuple)):
        return []
    seen, out = set(), []
    for raw in values:
        code = str(raw).strip().lower()
        if code in LANGUAGES and code not in seen:
            seen.add(code)
            out.append(code)
    return out


def languages_public() -> list[dict]:
    """The catalogue, for the onboarding picker."""
    return [
        {"code": code, "name": english, "endonym": endonym}
        for code, (english, endonym) in LANGUAGES.items()
    ]
